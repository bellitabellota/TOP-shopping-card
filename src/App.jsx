

import useFakeStoreAPI from "./helpers/useFakeStoreAPI";
import { useState } from "react";
import NavBar from "./sections/NavBar";
import { Outlet } from "react-router-dom";

function App() {
  const [selectedProducts, setSelectedProducts] = useState([]);

  const addProductToCart = (product) => {
    const productCard = document.querySelector(`[data-product-id="${product.id}"]`);
    const inputField = productCard.querySelector("input[type='number']");
    const quantityToAdd = Number(inputField.value);

    setSelectedProducts(prevProducts => {
      const index = prevProducts.findIndex((selectedProduct) =>  selectedProduct.id === product.id);
      const selectedProductsNew = prevProducts.map(product => ({...product}));

      if(index === -1) {
        selectedProductsNew.push({...product, quantity: quantityToAdd});
      } else {
        selectedProductsNew[index].quantity += quantityToAdd;
      }

      return selectedProductsNew;
    });
  }

  const { products, error, loading } = useFakeStoreAPI();
  
  if(loading) return <p>Products loading...</p>;
  if(error) return <p>{error.message}</p>;
  
  return (
    <>
      <NavBar selectedProducts={selectedProducts} />
      <Outlet context={[products, addProductToCart, selectedProducts]}/>
    </>

  )
}

export default App;