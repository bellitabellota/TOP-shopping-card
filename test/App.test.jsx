import { render, screen } from "@testing-library/react";
import App from "../src/App"
import { describe, it, vi, expect } from "vitest";

vi.mock("../src/helpers/useFakeStoreAPI.js", () => ({
  default: vi.fn(() => {
    return {product:[], error:null, loading:true }
  }),
}));

describe("App", ()=>{
  it("renders 'Products loading... when loading is true", () => {
    const {container} = render(<App />)
    expect(container).toMatchSnapshot();
  })
})