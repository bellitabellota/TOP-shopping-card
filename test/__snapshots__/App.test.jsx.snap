// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders 'Products loading... when loading is true 1`] = `
<div>
  <p>
    Products loading...
  </p>
</div>
`;
